databuilder/query_model_transforms.py

```diff
@@ -31,72 +36,49 @@ def apply_transforms(variables):
-    which reference it
-
-    This allows walking "backwards" along the graph which makes certain kinds of
-    transformation easier.
-    """
-    index = {}
-    for node in nodes:
-        populate_reverse_index(index, node)
-    return {node: references.values() for node, references in index.items()}
-
-
-def populate_reverse_index(index, node):
-    index.setdefault(node, {})
-    for subnode in get_input_nodes(node):
-        references = index.setdefault(subnode, {})
-        # Index nodes by their object ID so as to disambiguate distnct objects which
-        # compare equal
-        references[id(node)] = node
-        populate_reverse_index(index, subnode)
-
-
-def add_selected_columns_to_pick_row(reverse_index):
+def add_selected_columns_to_pick_row(nodes):
     """
     Replace instances of `PickOneRowPerPatient` with `PickOneRowPerPatientWithColumns`
     which track the columns that are going to be selected and allow us to generate the
     appropriate query
     """
-    for node, references in reverse_index.items():
+    for node in nodes:
         if not isinstance(node, PickOneRowPerPatient):
             continue
+
         # Get the name of any columns selected from this node
-        column_names = {c.name for c in references if isinstance(c, SelectColumn)}
+        dependers = get_dependers(node, nodes)
+        column_names = {c.name for c in dependers if isinstance(c, SelectColumn)}
+
         # Build a new set of SelectColumn operations which reference this node's source
         # instead
         selected_columns = frozenset(
             SelectColumn(node.source, name) for name in column_names
         )
-        # Create a new node which has the original attributes, plus the tuple of selected
-        # columns
-        new_node = PickOneRowPerPatientWithColumns(
-            source=node.source,
-            position=node.position,
-            selected_columns=selected_columns,
-        )
-        # Update any references to the old node to point to the new one
-        for ref_node in references:
-            update_references(ref_node, node, new_node)
 
+        # Modify the node in-place to have the new type
+        force_setattr(node, "__class__", PickOneRowPerPatientWithColumns)
+        force_setattr(node, "selected_columns", selected_columns)
 
-def update_references(target, old_ref, new_ref):
-    """
-    Replace references to `old_ref` with `new_ref` on `target`
-    """
-    # We're only expecting nodes with a source attribute here
-    assert target.source == old_ref
-    force_setattr(target, "source", new_ref)
-    # We're only expecting nodes which reference a single input
-    assert get_input_nodes(target) == [new_ref]
+
+def get_all_nodes(variables):
+    nodes = []
+    for query in variables.values():
+        nodes.extend(all_nodes(query))
+    return nodes
+
+
+def get_dependers(node, nodes):
+    dependers = []
+    for other_node in nodes:
```

Comment:
Given that this is called from inside a loop over `nodes`, this has quadratic behaviour.  Maybe ok for now?

Reply:
It does. I'm v happy to replace with a single pass and a dict if you prefer.

We have a _lot_ of QM graph traversal in the codebase. Could be worth doing some optimization at some point, but I'm not sure this is particularly like to be the worst offender.